### PR TITLE
Queryify `has_attr` to improve performance

### PR DESCRIPTION
**Pull request by camelid**
Inspired by https://github.com/rust-lang/rust/pull/127144#pullrequestreview-2149944996
and previous success in #94897.

r? @compiler-errors

### CODE REVIEW

File: compiler/rustc_middle/src/ty/util.rs [hunk 1]
```diff
@@ -1829,6 +1829,12 @@ pub fn reveal_opaque_types_in_bounds<'tcx>(
     val.fold_with(&mut visitor)
 }
 
+/// Determines whether an item is annotated with an attribute.
+pub fn has_attr_query(tcx: TyCtxt<'_>, key: (DefId, Symbol)) -> bool {
+    let (did, attr) = key;
```
Comment (fmease) [inline compiler/rustc_middle/src/ty/util.rs:1834]:
```suggestion
pub fn has_attr_query(tcx: TyCtxt<'_>, (did, attr): (DefId, Symbol)) -> bool {
```

Comment (camelid) [reply compiler/rustc_middle/src/ty/util.rs:1834]: Thanks, I'll change that after the perf run. I forgot to change it from the query declaration's signature :sweat_smile:

**Comment by camelid** [comment]
Hmm how could this cause such major regressions across the board? Is the extra overheard of the query system really that much?

**Comment by cjgillot** [comment]
Yes, that's the query overhead.

As a general rule, a query should either:
- take sufficiently long to make memorization + incr caching worth it;
- allow to hide enough information to avoid recomputing the caller query.

**Comment by camelid** [comment]
Makes sense, thanks!

**Pull request closed without merging**